The job **`x86_64-gnu-llvm-18`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/13009424266/job/36283579033) [(plain)](https://github.com/rust-lang/rust/commit/bb01b5a7d4ec3c5935a533cbd5522d9cd6ed64d1/checks/36283579033/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#21 exporting to docker image format
#21 sending tarball 31.8s done
#21 DONE 39.0s
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-18]
debug: `DISABLE_CI_RUSTC_IF_INCOMPATIBLE` configured.
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-18', '--enable-llvm-link-shared', '--set', 'rust.randomize-layout=true', '--set', 'rust.thin-lto-import-instr-limit=10', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-18/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.randomize-layout := True
configure: rust.thin-lto-import-instr-limit := 10
---
failures:

---- [codegen] tests/codegen/lib-optimizations/slice_rotate.rs stdout ----

error: verification with 'FileCheck' failed
status: exit status: 1
command: "/usr/lib/llvm-18/bin/FileCheck" "--input-file" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/lib-optimizations/slice_rotate/slice_rotate.ll" "/checkout/tests/codegen/lib-optimizations/slice_rotate.rs" "--check-prefix=CHECK" "--allow-unused-prefixes" "--dump-input-context" "100"
--- stderr -------------------------------
/checkout/tests/codegen/lib-optimizations/slice_rotate.rs:15:12: error: CHECK: expected string not found in input
/checkout/tests/codegen/lib-optimizations/slice_rotate.rs:15:12: error: CHECK: expected string not found in input
 // CHECK: %[[END:.+]].i = getelementptr inbounds i8, ptr %slice.0
           ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/lib-optimizations/slice_rotate/slice_rotate.ll:7:32: note: scanning from here
define void @rotate_left_by_one(ptr noalias nocapture noundef nonnull align 4 %slice.0, i64 noundef %slice.1) unnamed_addr #0 personality ptr @rust_eh_personality {
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/lib-optimizations/slice_rotate/slice_rotate.ll:18:2: note: possible intended match here
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/lib-optimizations/slice_rotate/slice_rotate.ll:18:2: note: possible intended match here
 %_10.i = getelementptr inbounds i32, ptr %slice.0, i64 1
/checkout/tests/codegen/lib-optimizations/slice_rotate.rs:46:12: error: CHECK: expected string not found in input
/checkout/tests/codegen/lib-optimizations/slice_rotate.rs:46:12: error: CHECK: expected string not found in input
 // CHECK: %dim.i.i.i = getelementptr inbounds i8, ptr %slice.0
           ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/lib-optimizations/slice_rotate/slice_rotate.ll:38:51: note: scanning from here
 %_10.i = getelementptr inbounds i32, ptr %slice.0, i64 %_7.0.i
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/lib-optimizations/slice_rotate/slice_rotate.ll:43:2: note: possible intended match here
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/lib-optimizations/slice_rotate/slice_rotate.ll:43:2: note: possible intended match here
 %2 = getelementptr inbounds i32, ptr %slice.0, i64 1

Input file: /checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/lib-optimizations/slice_rotate/slice_rotate.ll
Check file: /checkout/tests/codegen/lib-optimizations/slice_rotate.rs


-dump-input=help explains the following input dump.
Input was:
<<<<<<
<<<<<<
            1: ; ModuleID = 'slice_rotate.673fe9660e127e9f-cgu.0' 
            2: source_filename = "slice_rotate.673fe9660e127e9f-cgu.0" 
            3: target datalayout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-i128:128-f80:128-n8:16:32:64-S128" 
            5:  
            5:  
            6: ; Function Attrs: mustprogress nofree norecurse nosync nounwind nonlazybind willreturn memory(argmem: readwrite) uwtable 
            7: define void @rotate_left_by_one(ptr noalias nocapture noundef nonnull align 4 %slice.0, i64 noundef %slice.1) unnamed_addr #0 personality ptr @rust_eh_personality { 
check:15'0                                    X~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ error: no match found
check:15'0     ~~~~~~~
check:15'0     ~~~~~~~
            9:  %0 = icmp eq i64 %slice.1, 0 
check:15'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           10:  br i1 %0, label %bb3, label %bb2 
check:15'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
check:15'0     ~
check:15'0     ~
           12: bb2: ; preds = %start 
check:15'0     ~~~~~~~~~~~~~~~~~~~~~~
           13:  %_7.0.i = add i64 %slice.1, -1 
check:15'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           14:  %1 = icmp eq i64 %_7.0.i, 0 
check:15'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           15:  br i1 %1, label %bb3, label %bb5.i.i.i 
check:15'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
check:15'0     ~
check:15'0     ~
           17: bb5.i.i.i: ; preds = %bb2 
check:15'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~
           18:  %_10.i = getelementptr inbounds i32, ptr %slice.0, i64 1 
check:15'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
check:15'1      ?                                                         possible intended match
           19:  %2 = getelementptr inbounds i32, ptr %slice.0, i64 %_7.0.i 
check:15'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           20:  %3 = load i32, ptr %slice.0, align 4, !alias.scope !3 
check:15'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           21:  %4 = shl i64 %_7.0.i, 2 
check:15'0     ~~~~~~~~~~~~~~~~~~~~~~~~~
           22:  tail call void @llvm.memmove.p0.p0.i64(ptr nonnull align 4 %slice.0, ptr nonnull align 4 %_10.i, i64 %4, i1 false), !alias.scope !3 
check:15'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           23:  store i32 %3, ptr %2, align 4, !alias.scope !3 
check:15'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           24:  br label %bb3 
check:15'0     ~~~~~~~~~~~~~~~
check:15'0     ~
check:15'0     ~
           26: bb3: ; preds = %bb5.i.i.i, %bb2, %start 
check:15'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           27:  ret void 
check:15'0     ~~~~~~~~~~
           28: } 
check:15'0     ~~
check:15'0     ~
check:15'0     ~
           30: ; Function Attrs: mustprogress nofree norecurse nosync nounwind nonlazybind willreturn memory(argmem: readwrite) uwtable 
check:15'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           31: define void @rotate_right_by_one(ptr noalias nocapture noundef nonnull align 4 %slice.0, i64 noundef %slice.1) unnamed_addr #0 personality ptr @rust_eh_personality { 
check:15'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           32: start: 
           33:  %0 = icmp eq i64 %slice.1, 0 
           34:  br i1 %0, label %bb3, label %bb2 
           35:  
           36: bb2: ; preds = %start 
           37:  %_7.0.i = add i64 %slice.1, -1 
           38:  %_10.i = getelementptr inbounds i32, ptr %slice.0, i64 %_7.0.i 
check:46'0                                                       X~~~~~~~~~~~~~ error: no match found
           39:  %1 = icmp eq i64 %_7.0.i, 0 
check:46'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           40:  br i1 %1, label %bb3, label %bb2.i.i.i 
check:46'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
check:46'0     ~
check:46'0     ~
           42: bb2.i.i.i: ; preds = %bb2 
check:46'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~
           43:  %2 = getelementptr inbounds i32, ptr %slice.0, i64 1 
check:46'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
check:46'1      ?                                                     possible intended match
           44:  %.not.i.i.not.i = icmp eq i64 %_7.0.i, 1 
check:46'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           45:  br i1 %.not.i.i.not.i, label %bb3.i.i.i, label %bb4.i.i.i 
check:46'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
check:46'0     ~
check:46'0     ~
           47: bb4.i.i.i: ; preds = %bb2.i.i.i 
check:46'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           48:  %3 = load i32, ptr %_10.i, align 4, !alias.scope !6 
check:46'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           49:  %4 = shl i64 %_7.0.i, 2 
check:46'0     ~~~~~~~~~~~~~~~~~~~~~~~~~
           50:  tail call void @llvm.memmove.p0.p0.i64(ptr nonnull align 4 %2, ptr nonnull align 4 %slice.0, i64 %4, i1 false), !alias.scope !6 
check:46'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           51:  store i32 %3, ptr %slice.0, align 4, !alias.scope !6 
check:46'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           52:  br label %bb3 
check:46'0     ~~~~~~~~~~~~~~~
check:46'0     ~
check:46'0     ~
           54: bb3.i.i.i: ; preds = %bb2.i.i.i 
check:46'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           55:  %5 = load i32, ptr %slice.0, align 4, !alias.scope !6 
check:46'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           56:  %6 = load i32, ptr %_10.i, align 4, !alias.scope !6 
check:46'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           57:  store i32 %6, ptr %slice.0, align 4, !alias.scope !6 
check:46'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           58:  store i32 %5, ptr %2, align 4, !alias.scope !6 
check:46'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           59:  br label %bb3 
check:46'0     ~~~~~~~~~~~~~~~
check:46'0     ~
check:46'0     ~
           61: bb3: ; preds = %bb3.i.i.i, %bb4.i.i.i, %bb2, %start 
check:46'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           62:  ret void 
check:46'0     ~~~~~~~~~~
           63: } 
check:46'0     ~~
check:46'0     ~
check:46'0     ~
           65: ; Function Attrs: nounwind nonlazybind uwtable 
check:46'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           66: declare noundef i32 @rust_eh_personality(i32 noundef, i32 noundef, i64 noundef, ptr noundef, ptr noundef) unnamed_addr #1 
check:46'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
check:46'0     ~
check:46'0     ~
           68: ; Function Attrs: mustprogress nocallback nofree nounwind willreturn memory(argmem: readwrite) 
check:46'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           69: declare void @llvm.memmove.p0.p0.i64(ptr nocapture writeonly, ptr nocapture readonly, i64, i1 immarg) #2 
check:46'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
check:46'0     ~
check:46'0     ~
           71: attributes #0 = { mustprogress nofree norecurse nosync nounwind nonlazybind willreturn memory(argmem: readwrite) uwtable "probe-stack"="inline-asm" "target-cpu"="x86-64" "target-features"="+x87,+sse2" } 
check:46'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           72: attributes #1 = { nounwind nonlazybind uwtable "probe-stack"="inline-asm" "target-cpu"="x86-64" "target-features"="+x87,+sse2" } 
check:46'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           73: attributes #2 = { mustprogress nocallback nofree nounwind willreturn memory(argmem: readwrite) } 
check:46'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
check:46'0     ~
check:46'0     ~
           75: !llvm.module.flags = !{!0, !1} 
check:46'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           76: !llvm.ident = !{!2} 
check:46'0     ~~~~~~~~~~~~~~~~~~~~
check:46'0     ~
check:46'0     ~
           78: !0 = !{i32 8, !"PIC Level", i32 2} 
check:46'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           79: !1 = !{i32 2, !"RtLibUseGOT", i32 1} 
check:46'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           80: !2 = !{!"rustc version 1.86.0-nightly (d7cc7b341 2025-01-28)"} 
check:46'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           81: !3 = !{!4} 
check:46'0     ~~~~~~~~~~~
           82: !4 = distinct !{!4, !5, !"_ZN4core5slice29_$LT$impl$u20$$u5b$T$u5d$$GT$11rotate_left17h7976afab32dcf97dE: %self.0"} 
check:46'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           83: !5 = distinct !{!5, !"_ZN4core5slice29_$LT$impl$u20$$u5b$T$u5d$$GT$11rotate_left17h7976afab32dcf97dE"} 
check:46'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           84: !6 = !{!7} 
check:46'0     ~~~~~~~~~~~
           85: !7 = distinct !{!7, !8, !"_ZN4core5slice29_$LT$impl$u20$$u5b$T$u5d$$GT$12rotate_right17h291a3c4ebe40cb8aE: %self.0"} 
check:46'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           86: !8 = distinct !{!8, !"_ZN4core5slice29_$LT$impl$u20$$u5b$T$u5d$$GT$12rotate_right17h291a3c4ebe40cb8aE"} 
check:46'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
------------------------------------------



```

</details>

--- a/library/core/src/slice/rotate.rs
+++ b/library/core/src/slice/rotate.rs
@@ -1,6 +1,8 @@
 use crate::mem::{self, MaybeUninit, SizedTypeProperties};
 use crate::{cmp, ptr};
 
+type BufType = [usize; 32];
+
 /// Rotates the range `[mid-left, mid+right)` such that the element at `mid` becomes the first
 /// element. Equivalently, rotates the range `left` elements to the left or `right` elements to the
 /// right.
@@ -8,17 +10,76 @@ use crate::{cmp, ptr};
 /// # Safety
 ///
 /// The specified range must be valid for reading and writing.
-///
-/// # Algorithm
-///
+pub(super) unsafe fn ptr_rotate<T>(left: usize, mid: *mut T, right: usize) {
+    if T::IS_ZST {
+        return;
+    }
+    // abort early if the rotate is a no-op
+    if (left == 0) || (right == 0) {
+        return;
+    }
+    // `T` is not a zero-sized type, so it's okay to divide by its size.
+    if !cfg!(feature = "optimize_for_size")
+        && cmp::min(left, right) <= mem::size_of::<BufType>() / mem::size_of::<T>()
+    {
+        // SAFETY: guaranteed by the caller
+        unsafe { ptr_rotate_memmove(left, mid, right) };
+    } else if !cfg!(feature = "optimize_for_size")
+        && ((left + right < 24) || (mem::size_of::<T>() > mem::size_of::<[usize; 4]>()))
+    {
+        // SAFETY: guaranteed by the caller
+        unsafe { ptr_rotate_gcd(left, mid, right) }
+    } else {
+        // SAFETY: guaranteed by the caller
+        unsafe { ptr_rotate_swap(left, mid, right) }
+    }
+}
+
 /// Algorithm 1 is used if `min(left, right)` is small enough to fit onto a stack buffer. The
 /// `min(left, right)` elements are copied onto the buffer, `memmove` is applied to the others, and
 /// the ones on the buffer are moved back into the hole on the opposite side of where they
 /// originated.
 ///
-/// Algorithms that can be vectorized outperform the above once `left + right` becomes large enough.
+/// # Safety
 ///
-/// Algorithm 2 is otherwise used for small values of `left + right` or for large `T`. The elements
+/// The specified range must be valid for reading and writing.
+unsafe fn ptr_rotate_memmove<T>(left: usize, mid: *mut T, right: usize) {
+    // The `[T; 0]` here is to ensure this is appropriately aligned for T
+    let mut rawarray = MaybeUninit::<(BufType, [T; 0])>::uninit();
+    let buf = rawarray.as_mut_ptr() as *mut T;
+    // SAFETY: `mid-left <= mid-left+right < mid+right`
+    let dim = unsafe { mid.sub(left).add(right) };
+    if left <= right {
+        // SAFETY:
+        //
+        // 1) The `if` condition about the sizes ensures `[mid-left; left]` will fit in
+        //    `buf` without overflow and `buf` was created just above and so cannot be
+        //    overlapped with any value of `[mid-left; left]`
+        // 2) [mid-left, mid+right) are all valid for reading and writing and we don't care
+        //    about overlaps here.
+        // 3) The `if` condition about `left <= right` ensures writing `left` elements to
+        //    `dim = mid-left+right` is valid because:
+        //    - `buf` is valid and `left` elements were written in it in 1)
+        //    - `dim+left = mid-left+right+left = mid+right` and we write `[dim, dim+left)`
+        unsafe {
+            // 1)
+            ptr::copy_nonoverlapping(mid.sub(left), buf, left);
+            // 2)
+            ptr::copy(mid, mid.sub(left), right);
+            // 3)
+            ptr::copy_nonoverlapping(buf, dim, left);
+        }
+    } else {
+        // SAFETY: same reasoning as above but with `left` and `right` reversed
+        unsafe {
+            ptr::copy_nonoverlapping(mid, buf, right);
+            ptr::copy(mid.sub(left), dim, left);
+            ptr::copy_nonoverlapping(buf, mid.sub(left), right);
+        }
+    }
+}
+
+/// Algorithm 2 is used for small values of `left + right` or for large `T`. The elements
 /// are moved into their final positions one at a time starting at `mid - left` and advancing by
 /// `right` steps modulo `left + right`, such that only one temporary is needed. Eventually, we
 /// arrive back at `mid - left`. However, if `gcd(left + right, right)` is not 1, the above steps
@@ -48,9 +109,101 @@ use crate::{cmp, ptr};
 ///
 /// Algorithm 2 can be vectorized by chunking and performing many rounds at once, but there are too
 /// few rounds on average until `left + right` is enormous, and the worst case of a single
-/// round is always there. Instead, algorithm 3 utilizes repeated swapping of
-/// `min(left, right)` elements until a smaller rotate problem is left.
+/// round is always there.
 ///
+/// # Safety
+///
+/// The specified range must be valid for reading and writing.
+unsafe fn ptr_rotate_gcd<T>(left: usize, mid: *mut T, right: usize) {
+    // Algorithm 2
+    // Microbenchmarks indicate that the average performance for random shifts is better all
+    // the way until about `left + right == 32`, but the worst case performance breaks even
+    // around 16. 24 was chosen as middle ground. If the size of `T` is larger than 4
+    // `usize`s, this algorithm also outperforms other algorithms.
+    // SAFETY: callers must ensure `mid - left` is valid for reading and writing.
+    let x = unsafe { mid.sub(left) };
+    // beginning of first round
+    // SAFETY: see previous comment.
+    let mut tmp: T = unsafe { x.read() };
+    let mut i = right;
+    // `gcd` can be found before hand by calculating `gcd(left + right, right)`,
+    // but it is faster to do one loop which calculates the gcd as a side effect, then
+    // doing the rest of the chunk
+    let mut gcd = right;
+    // benchmarks reveal that it is faster to swap temporaries all the way through instead
+    // of reading one temporary once, copying backwards, and then writing that temporary at
+    // the very end. This is possibly due to the fact that swapping or replacing temporaries
+    // uses only one memory address in the loop instead of needing to manage two.
+    loop {
+        // [long-safety-expl]
+        // SAFETY: callers must ensure `[left, left+mid+right)` are all valid for reading and
+        // writing.
+        //
+        // - `i` start with `right` so `mid-left <= x+i = x+right = mid-left+right < mid+right`
+        // - `i <= left+right-1` is always true
+        //   - if `i < left`, `right` is added so `i < left+right` and on the next
+        //     iteration `left` is removed from `i` so it doesn't go further
+        //   - if `i >= left`, `left` is removed immediately and so it doesn't go further.
+        // - overflows cannot happen for `i` since the function's safety contract ask for
+        //   `mid+right-1 = x+left+right` to be valid for writing
+        // - underflows cannot happen because `i` must be bigger or equal to `left` for
+        //   a subtraction of `left` to happen.
+        //
+        // So `x+i` is valid for reading and writing if the caller respected the contract
+        tmp = unsafe { x.add(i).replace(tmp) };
+        // instead of incrementing `i` and then checking if it is outside the bounds, we
+        // check if `i` will go outside the bounds on the next increment. This prevents
+        // any wrapping of pointers or `usize`.
+        if i >= left {
+            i -= left;
+            if i == 0 {
+                // end of first round
+                // SAFETY: tmp has been read from a valid source and x is valid for writing
+                // according to the caller.
+                unsafe { x.write(tmp) };
+                break;
+            }
+            // this conditional must be here if `left + right >= 15`
+            if i < gcd {
+                gcd = i;
+            }
+        } else {
+            i += right;
+        }
+    }
+    // finish the chunk with more rounds
+    for start in 1..gcd {
+        // SAFETY: `gcd` is at most equal to `right` so all values in `1..gcd` are valid for
+        // reading and writing as per the function's safety contract, see [long-safety-expl]
+        // above
+        tmp = unsafe { x.add(start).read() };
+        // [safety-expl-addition]
+        //
+        // Here `start < gcd` so `start < right` so `i < right+right`: `right` being the
+        // greatest common divisor of `(left+right, right)` means that `left = right` so
+        // `i < left+right` so `x+i = mid-left+i` is always valid for reading and writing
+        // according to the function's safety contract.
+        i = start + right;
+        loop {
+            // SAFETY: see [long-safety-expl] and [safety-expl-addition]
+            tmp = unsafe { x.add(i).replace(tmp) };
+            if i >= left {
+                i -= left;
+                if i == start {
+                    // SAFETY: see [long-safety-expl] and [safety-expl-addition]
+                    unsafe { x.add(start).write(tmp) };
+                    break;
+                }
+            } else {
+                i += right;
+            }
+        }
+    }
+}
+
+/// Algorithm 3 utilizes repeated swapping of `min(left, right)` elements.
+///
+/// ///
 /// ```text
 /// left = 11, right = 4
 /// [4 5 6 7 8 9 10 11 12 13 14 . 0 1 2 3]
@@ -61,182 +214,50 @@ use crate::{cmp, ptr};
 /// we cannot swap any more, but a smaller rotation problem is left to solve
 /// ```
 /// when `left < right` the swapping happens from the left instead.
-pub(super) unsafe fn ptr_rotate<T>(mut left: usize, mut mid: *mut T, mut right: usize) {
-    type BufType = [usize; 32];
-    if T::IS_ZST {
-        return;
-    }
-    // N.B. the below algorithms can fail if these cases are not checked
-    if (right == 0) || (left == 0) {
-        return;
-    }
-    // `T` is not a zero-sized type, so it's okay to divide by its size.
-    if !cfg!(feature = "optimize_for_size")
-        && cmp::min(left, right) <= mem::size_of::<BufType>() / mem::size_of::<T>()
-    {
-        // Algorithm 1
-        // The `[T; 0]` here is to ensure this is appropriately aligned for T
-        let mut rawarray = MaybeUninit::<(BufType, [T; 0])>::uninit();
-        let buf = rawarray.as_mut_ptr() as *mut T;
-        // SAFETY: `mid-left <= mid-left+right < mid+right`
-        let dim = unsafe { mid.sub(left).add(right) };
-        if left <= right {
-            // SAFETY:
-            //
-            // 1) The `if` condition about the sizes ensures `[mid-left; left]` will fit in
-            //    `buf` without overflow and `buf` was created just above and so cannot be
-            //    overlapped with any value of `[mid-left; left]`
-            // 2) [mid-left, mid+right) are all valid for reading and writing and we don't care
-            //    about overlaps here.
-            // 3) The `if` condition about `left <= right` ensures writing `left` elements to
-            //    `dim = mid-left+right` is valid because:
-            //    - `buf` is valid and `left` elements were written in it in 1)
-            //    - `dim+left = mid-left+right+left = mid+right` and we write `[dim, dim+left)`
-            unsafe {
-                // 1)
-                ptr::copy_nonoverlapping(mid.sub(left), buf, left);
-                // 2)
-                ptr::copy(mid, mid.sub(left), right);
-                // 3)
-                ptr::copy_nonoverlapping(buf, dim, left);
-            }
-        } else {
-            // SAFETY: same reasoning as above but with `left` and `right` reversed
-            unsafe {
-                ptr::copy_nonoverlapping(mid, buf, right);
-                ptr::copy(mid.sub(left), dim, left);
-                ptr::copy_nonoverlapping(buf, mid.sub(left), right);
-            }
-        }
-    } else if !cfg!(feature = "optimize_for_size")
-        && ((left + right < 24) || (mem::size_of::<T>() > mem::size_of::<[usize; 4]>()))
-    {
-        // Algorithm 2
-        // Microbenchmarks indicate that the average performance for random shifts is better all
-        // the way until about `left + right == 32`, but the worst case performance breaks even
-        // around 16. 24 was chosen as middle ground. If the size of `T` is larger than 4
-        // `usize`s, this algorithm also outperforms other algorithms.
-        // SAFETY: callers must ensure `mid - left` is valid for reading and writing.
-        let x = unsafe { mid.sub(left) };
-        // beginning of first round
-        // SAFETY: see previous comment.
-        let mut tmp: T = unsafe { x.read() };
-        let mut i = right;
-        // `gcd` can be found before hand by calculating `gcd(left + right, right)`,
-        // but it is faster to do one loop which calculates the gcd as a side effect, then
-        // doing the rest of the chunk
-        let mut gcd = right;
-        // benchmarks reveal that it is faster to swap temporaries all the way through instead
-        // of reading one temporary once, copying backwards, and then writing that temporary at
-        // the very end. This is possibly due to the fact that swapping or replacing temporaries
-        // uses only one memory address in the loop instead of needing to manage two.
-        loop {
-            // [long-safety-expl]
-            // SAFETY: callers must ensure `[left, left+mid+right)` are all valid for reading and
-            // writing.
-            //
-            // - `i` start with `right` so `mid-left <= x+i = x+right = mid-left+right < mid+right`
-            // - `i <= left+right-1` is always true
-            //   - if `i < left`, `right` is added so `i < left+right` and on the next
-            //     iteration `left` is removed from `i` so it doesn't go further
-            //   - if `i >= left`, `left` is removed immediately and so it doesn't go further.
-            // - overflows cannot happen for `i` since the function's safety contract ask for
-            //   `mid+right-1 = x+left+right` to be valid for writing
-            // - underflows cannot happen because `i` must be bigger or equal to `left` for
-            //   a subtraction of `left` to happen.
-            //
-            // So `x+i` is valid for reading and writing if the caller respected the contract
-            tmp = unsafe { x.add(i).replace(tmp) };
-            // instead of incrementing `i` and then checking if it is outside the bounds, we
-            // check if `i` will go outside the bounds on the next increment. This prevents
-            // any wrapping of pointers or `usize`.
-            if i >= left {
-                i -= left;
-                if i == 0 {
-                    // end of first round
-                    // SAFETY: tmp has been read from a valid source and x is valid for writing
-                    // according to the caller.
-                    unsafe { x.write(tmp) };
+///
+/// # Safety
+///
+/// The specified range must be valid for reading and writing.
+unsafe fn ptr_rotate_swap<T>(mut left: usize, mut mid: *mut T, mut right: usize) {
+    loop {
+        if left >= right {
+            // Algorithm 3
+            // There is an alternate way of swapping that involves finding where the last swap
+            // of this algorithm would be, and swapping using that last chunk instead of swapping
+            // adjacent chunks like this algorithm is doing, but this way is still faster.
+            loop {
+                // SAFETY:
+                // `left >= right` so `[mid-right, mid+right)` is valid for reading and writing
+                // Subtracting `right` from `mid` each turn is counterbalanced by the addition and
+                // check after it.
+                unsafe {
+                    ptr::swap_nonoverlapping(mid.sub(right), mid, right);
+                    mid = mid.sub(right);
+                }
+                left -= right;
+                if left < right {
                     break;
                 }
-                // this conditional must be here if `left + right >= 15`
-                if i < gcd {
-                    gcd = i;
-                }
-            } else {
-                i += right;
             }
-        }
-        // finish the chunk with more rounds
-        for start in 1..gcd {
-            // SAFETY: `gcd` is at most equal to `right` so all values in `1..gcd` are valid for
-            // reading and writing as per the function's safety contract, see [long-safety-expl]
-            // above
-            tmp = unsafe { x.add(start).read() };
-            // [safety-expl-addition]
-            //
-            // Here `start < gcd` so `start < right` so `i < right+right`: `right` being the
-            // greatest common divisor of `(left+right, right)` means that `left = right` so
-            // `i < left+right` so `x+i = mid-left+i` is always valid for reading and writing
-            // according to the function's safety contract.
-            i = start + right;
+        } else {
+            // Algorithm 3, `left < right`
             loop {
-                // SAFETY: see [long-safety-expl] and [safety-expl-addition]
-                tmp = unsafe { x.add(i).replace(tmp) };
-                if i >= left {
-                    i -= left;
-                    if i == start {
-                        // SAFETY: see [long-safety-expl] and [safety-expl-addition]
-                        unsafe { x.add(start).write(tmp) };
-                        break;
-                    }
-                } else {
-                    i += right;
+                // SAFETY: `[mid-left, mid+left)` is valid for reading and writing because
+                // `left < right` so `mid+left < mid+right`.
+                // Adding `left` to `mid` each turn is counterbalanced by the subtraction and check
+                // after it.
+                unsafe {
+                    ptr::swap_nonoverlapping(mid.sub(left), mid, left);
+                    mid = mid.add(left);
+                }
+                right -= left;
+                if right < left {
+                    break;
                 }
             }
         }
-    } else {
-        loop {
-            if left >= right {
-                // Algorithm 3
-                // There is an alternate way of swapping that involves finding where the last swap
-                // of this algorithm would be, and swapping using that last chunk instead of swapping
-                // adjacent chunks like this algorithm is doing, but this way is still faster.
-                loop {
-                    // SAFETY:
-                    // `left >= right` so `[mid-right, mid+right)` is valid for reading and writing
-                    // Subtracting `right` from `mid` each turn is counterbalanced by the addition and
-                    // check after it.
-                    unsafe {
-                        ptr::swap_nonoverlapping(mid.sub(right), mid, right);
-                        mid = mid.sub(right);
-                    }
-                    left -= right;
-                    if left < right {
-                        break;
-                    }
-                }
-            } else {
-                // Algorithm 3, `left < right`
-                loop {
-                    // SAFETY: `[mid-left, mid+left)` is valid for reading and writing because
-                    // `left < right` so `mid+left < mid+right`.
-                    // Adding `left` to `mid` each turn is counterbalanced by the subtraction and check
-                    // after it.
-                    unsafe {
-                        ptr::swap_nonoverlapping(mid.sub(left), mid, left);
-                        mid = mid.add(left);
-                    }
-                    right -= left;
-                    if right < left {
-                        break;
-                    }
-                }
-            }
-
-            if (right == 0) || (left == 0) {
-                return;
-            }
+        if (right == 0) || (left == 0) {
+            return;
         }
     }
 }

--- a/library/core/src/slice/rotate.rs
+++ b/library/core/src/slice/rotate.rs
@@ -10,6 +10,7 @@ type BufType = [usize; 32];
 /// # Safety
 ///
 /// The specified range must be valid for reading and writing.
+#[inline]
 pub(super) unsafe fn ptr_rotate<T>(left: usize, mid: *mut T, right: usize) {
     if T::IS_ZST {
         return;
@@ -43,6 +44,7 @@ pub(super) unsafe fn ptr_rotate<T>(left: usize, mid: *mut T, right: usize) {
 /// # Safety
 ///
 /// The specified range must be valid for reading and writing.
+#[inline]
 unsafe fn ptr_rotate_memmove<T>(left: usize, mid: *mut T, right: usize) {
     // The `[T; 0]` here is to ensure this is appropriately aligned for T
     let mut rawarray = MaybeUninit::<(BufType, [T; 0])>::uninit();
@@ -114,6 +116,7 @@ unsafe fn ptr_rotate_memmove<T>(left: usize, mid: *mut T, right: usize) {
 /// # Safety
 ///
 /// The specified range must be valid for reading and writing.
+#[inline]
 unsafe fn ptr_rotate_gcd<T>(left: usize, mid: *mut T, right: usize) {
     // Algorithm 2
     // Microbenchmarks indicate that the average performance for random shifts is better all
@@ -218,6 +221,7 @@ unsafe fn ptr_rotate_gcd<T>(left: usize, mid: *mut T, right: usize) {
 /// # Safety
 ///
 /// The specified range must be valid for reading and writing.
+#[inline]
 unsafe fn ptr_rotate_swap<T>(mut left: usize, mut mid: *mut T, mut right: usize) {
     loop {
         if left >= right {

--- a/tests/codegen/lib-optimizations/slice_rotate.rs
+++ b/tests/codegen/lib-optimizations/slice_rotate.rs
@@ -1,0 +1,30 @@
+//@ compile-flags: -O
+
+#![crate_type = "lib"]
+
+// Ensure that the simple case of rotating by a constant 1 optimizes to the obvious thing
+
+// CHECK-LABEL: @rotate_left_by_one
+#[no_mangle]
+pub fn rotate_left_by_one(slice: &mut [i32]) {
+    // CHECK-NOT: phi
+    // CHECK-NOT: call
+    // CHECK-NOT: load
+    // CHECK-NOT: store
+    // CHECK-NOT: getelementptr
+    // CHECK: %[[END:.+]] = getelementptr
+    // CHECK-NEXT: %[[DIM:.+]] = getelementptr
+    // CHECK-NEXT: %[[LAST:.+]] = load
+    // CHECK-NEXT: %[[FIRST:.+]] = shl
+    // CHECK-NEXT: call void @llvm.memmove
+    // CHECK-NEXT: store i32 %[[LAST]], ptr %[[DIM:.+]]
+    // CHECK-NOT: phi
+    // CHECK-NOT: call
+    // CHECK-NOT: load
+    // CHECK-NOT: store
+    // CHECK-NOT: getelementptr
+    // CHECK: ret void
+    if !slice.is_empty() {
+        slice.rotate_left(1);
+    }
+}

